Add LOOP-compatible Head Reporter

reductionista opened the pull request:
<!--- Does this work have a corresponding ticket?  Please link it here as well as one of:
    - the PR title
    - branch name
    - commit message
[NONEVM-1642](https://app.naviabenefits.com/#/login)
--> 

# Description

Presently, we only have an evm implementation (`headreporter.telemetryReporter`) of `HeadReporter` based on the legacy EVM chains.

This re-names `headreporter.telemetryReporter` to `headreporter.evmTelemetryReporter` and adds `headreporter.loopTelemetryReporter`, a chain-agnostic implementation of `HeadReporter`.

This should work for any chains that fully implement the Relayer interface (specifically LatestHead) since they are all required to satisfy the same interface. We don't want to replace the legacy EVM implementation yet with it for EVM, because that would remove reporting on finalized heads.

Adding support for finalized heads later will require extending the relayer interface. To avoid removing any functionality from EVM, this will run side-by-side the legacy EVM implementation, which can be removed once we add LatestFinalizedHead (or similar) to the relayer interface. There will be some extra reporting on latest heads for evm now, but these are cached in memory so it shouldn't add any overhead.

[NONEVM-1642]: https://smartcontract-it.atlassian.net/browse/NONEVM-1642?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

reductionista committed:
Add Solana Head Reporter

DylanTinianov commented on core/services/chainlink/application.go:
```diff
@@ -472,12 +472,14 @@ func NewApplication(ctx context.Context, opts ApplicationOpts) (Application, err
 	}
-	telemReporter := headreporter.NewTelemetryReporter(telemetryManager, globalLogger, chainIDs...)
-	headReporter := headreporter.NewHeadReporterService(opts.DS, globalLogger, promReporter, telemReporter)
+	solanaRelayers, _ := relayChainInterops.List(FilterRelayersByType("Solana")).GetIDToRelayerMap()
```
Yeah we should probably handle this error here

reductionista replied:
It seemed a bit silly to handle it when the function always returns nil. I guess I'll just update the signature so that it doesn't look like it can return an error.

prashantkumar1982 commented on core/services/chainlink/application.go:
```diff
@@ -472,12 +472,14 @@ func NewApplication(ctx context.Context, opts ApplicationOpts) (Application, err
 	srvcs = append(srvcs, workflowORM)
 
 	promReporter := headreporter.NewPrometheusReporter(opts.DS, legacyEVMChains)
```
While you're at this, could you also rename this to evmPromReporter in all places, just to make it clear that even this prom reporter is only EVM specific.

prashantkumar1982 replied:
Actually legacyEvmPromReporter

prashantkumar1982 commented on core/services/chainlink/application.go:
```diff
@@ -472,12 +472,14 @@ func NewApplication(ctx context.Context, opts ApplicationOpts) (Application, err
 	}
-	telemReporter := headreporter.NewTelemetryReporter(telemetryManager, globalLogger, chainIDs...)
-	headReporter := headreporter.NewHeadReporterService(opts.DS, globalLogger, promReporter, telemReporter)
+	solanaRelayers, _ := relayChainInterops.List(FilterRelayersByType("Solana")).GetIDToRelayerMap()
```
Don't use the literal "solana".
Use relay.NetworkSolana var.

prashantkumar1982 commented on core/services/headreporter/telemetry_reporter.go:
```diff
@@ -2,59 +2,65 @@ package headreporter
 }
 
-func NewTelemetryReporter(monitoringEndpointGen telemetry.MonitoringEndpointGenerator, lggr logger.Logger, chainIDs ...*big.Int) HeadReporter {
+func NewEVMTelemetryReporter(monitoringEndpointGen telemetry.MonitoringEndpointGenerator, lggr logger.Logger, chainIDs ...*big.Int) HeadReporter {
```
Thanks for renaming!
Could you please rename to legacyEVM at all places?
So that we're clear that this is using legacy EVM flow

prashantkumar1982 commented on core/services/headreporter/telemetry_reporter.go:
```diff
@@ -2,59 +2,65 @@ package headreporter
 		finalized = &telem.Block{
-			Timestamp: uint64(latestFinalizedHead.GetTimestamp().UTC().Unix()),
-			Number:    uint64(latestFinalizedHead.BlockNumber()),
+			Timestamp: utils.NonNegativeInt64ToUint64(latestFinalizedHead.GetTimestamp().UTC().Unix()), // golint:gosec
```
Why do you still need golint:gosec in these lines?

reductionista replied:
We don't, missed that

prashantkumar1982 commented on core/services/headreporter/telemetry_reporter.go:
```diff
@@ -64,6 +70,76 @@ func (t *telemetryReporter) ReportNewHead(ctx context.Context, head *evmtypes.He
+}
+
+// ReportNewHead is unimplemented on Solana because there is no Headtracker to subscribe to
+func (t *solanaTelemetryReporter) ReportNewHead(_ context.Context, _ *evmtypes.Head) error {
```
Don't like the fact that this method still needs an evmtypes.Head, even though this is part of the interface for Solana.
Seems easy to change to a generic head, but perhaps lets do it in a separate PR later.

reductionista replied:
Yeah, I considered making that change, but I think this `ReportNewHead` method will need to be removed soon anyway. The only reason EVM is able to use `ReportNewHead` (registered callback) rather than `PeriodicReport` (polling) is because it's not running in LOOP. I don'tt hink non-evm chains would ever be able to use it unless we changed the way our LOOP arch is set up so that relayers could make callbacks to the core node.

prashantkumar1982 commented on core/services/headreporter/telemetry_reporter.go:
```diff
@@ -64,6 +70,76 @@ func (t *telemetryReporter) ReportNewHead(ctx context.Context, head *evmtypes.He
+	relays := make(map[string]loop.Relayer)
+	for relayID, relay := range solanaRelays {
+		chainID := relayID.ChainID
+		endpoints[relayID.ChainID] = monitoringEndpointGen.GenMonitoringEndpoint("Solana", chainID, "", synchronization.HeadReport)
```
rename "Solana" with relay.NetworkSolana

prashantkumar1982 commented on core/services/headreporter/telemetry_reporter.go:
```diff
@@ -64,6 +70,76 @@ func (t *telemetryReporter) ReportNewHead(ctx context.Context, head *evmtypes.He
+	return nil
+}
+
+type solanaTelemetryReporter struct {
```
nice work!
I see there's literally nothing solana specific here.
What do you think about renaming it to genericTelemReporter, and enable it for all chain families, perhaps except EVM, or maybe don't even need to filter out EVM.
This will just add more telem reporting of EVM Heads at periodic intervals too. I already checked, in EVM relayer's LatestHead(), we return a cached version of head, so no extra RPCs are made.

No other non-evm is launched on CCIP yet, so this change would be no-op in production till they launch. 
I only see Aptos and Tron DF being live that would start reporting these Heads, but not a bad thing.
What do you think?

reductionista replied:
Yeah, I guess we could do that.

I was mostly thinking that the amount of QA testing needed might be a lot less if we know the change only affects Solana. But you have a good point, I don't think `LatestHead` is implemented for any families but EVM & Solana so the other chains shouldn't matter.  And if it's cached on EVM it shouldn't make much of a difference there and can be tested simultaneously with Solana on staging.

prashantkumar1982 commented on core/utils/safe_math.go:
```diff
@@ -0,0 +1,9 @@
+	if i < 0 {
+		i = 0
+	}
+	return uint64(i) //nolint:gosec
```
Did you get any benefit by creating this, then, if you still get this lint warning?
I would say, since we know we are converting from a positive number, lets just ignore this warning

reductionista replied:
The benefit was that, there were at least 6 different lines where I had to add `//nolint:gosec` to get CI to pass... now it's contained to just a single location.  Several times I thought I'd gotten all of them and then the CI would fail again.

reductionista replied:
Next time I'm making an update to chainlink-common I'd like to move it there, and re-use it also in LogPoller since that is also filled with a lot of the same //nolint:gosec's

reductionista replied:
Also, it seems slightly safer in that... if somehow there were a situation where a negative block number were passed it wouldn't accidentally get converted into a positive number. And if there were a clock that was set prior to 1970 it would cause less unpredictable behavior.

reductionista committed:
Add explanation to //golint:nosec

reductionista committed:
Address PR comments

Renames:
- NewEVMTelemetry -> NewLegacyEVMTelemetry
- NewPrometheusReporter -> NewLegacyEVMPrometheusReporter
- evmTelemetryReporter -> legacyEVMTelemetryReporter
- NewSolanaTelemetryReporter -> NewTelemetryReporter
- solanaTelemetryReporter -> loopTelemetryReporter

Now, NewTelemetryReporter handles all relayers in the same way
(including EVM). The legacy EVM Telemetry Reporter is left operational
so we still receive metrics about finalized heads. This can be removed
once we extend the relayer interface to be able to fetch finalized heads
from there.

reductionista committed:
Remove unused error output from GetIDToRelayerMap()
